src/main/java/com/coveo/saml/SamlClient.java

```diff
@@ -548,34 +551,28 @@ private static DOMParser createDOMParser() throws SamlException {
     return parser;
   }
 
-  private static MetadataProvider createMetadataProvider(Reader metadata) throws SamlException {
+  private static DOMMetadataResolver createMetadataResolver(Reader metadata) throws SamlException {
     try {
       DOMParser parser = createDOMParser();
       parser.parse(new InputSource(metadata));
-      DOMMetadataProvider provider =
-          new DOMMetadataProvider(parser.getDocument().getDocumentElement());
-      provider.initialize();
-      return provider;
-    } catch (IOException | SAXException | MetadataProviderException ex) {
+      DOMMetadataResolver resolver =
+          new DOMMetadataResolver(parser.getDocument().getDocumentElement());
+      resolver.setId("componentId"); // The resolver needs an ID for the initialization to go through.
+      resolver.initialize();
+      return resolver;
+    } catch (IOException | SAXException | ComponentInitializationException ex) {
       throw new SamlException("Cannot load identity provider metadata", ex);
     }
   }
 
-  private static EntityDescriptor getEntityDescriptor(MetadataProvider metadataProvider)
+  private static EntityDescriptor getEntityDescriptor(DOMMetadataResolver metadata)
       throws SamlException {
-    EntityDescriptor descriptor;
-
-    try {
-      descriptor = (EntityDescriptor) metadataProvider.getMetadata();
-    } catch (MetadataProviderException ex) {
-      throw new SamlException("Cannot retrieve the entity descriptor", ex);
+    List<EntityDescriptor> entityDescriptors = new ArrayList<>();
+    metadata.forEach(entityDescriptors::add);
+    if (entityDescriptors.size() != 1) {
```

Comment:
pretty sure that there are cases where it makes sense to have more than one, but I'd have to do a bit of reading to confirm...